### PR TITLE
SIMD for Silk.NET.Math

### PR DESCRIPTION
**Pull request by WhiteBlackGoose**
## Summary of the PR
Starting implementing the [SIMD proposal](https://github.com/dotnet/Silk.NET/blob/main/documentation/proposals/Proposal%20-%20Vectorization%20-%20SIMD.md).

## Plan for this PR

All methods are for three bitnesses: 64, 128, 256.

- [x] `Create`
- [x] `Load`
- [x] `Add`
- [x] `Subtract`
- [x] `Multiply`
- [x] `Divide`
- [x] `Abs`
- [x] Constants
- [x] `And`
- [x] `Or`
- [x] `Xor`
- [x] `Not`
- [x] `AllBitsSet`
- [x] `IsHardwareAccelerated`

Other:
- [x] Docs

### CODE REVIEW

File: src/Maths/Silk.NET.Maths/Simd/Simd128.Abs.cs
Comment (HurricanKai) [inline src/Maths/Silk.NET.Maths/Simd/Simd128.Abs.cs:129]: I think just calling down to Simd128.And is preferable here?

File: src/Maths/Silk.NET.Maths/Simd/Simd128.Add.cs
Comment (HurricanKai) [inline src/Maths/Silk.NET.Maths/Simd/Simd128.Add.cs:49]: This doesn't make sense - if AVX2 is supported, SSE2 is as well, so the lower branch is never hit - but I'm unsure whether we should prefer SSE or AVX here
Edit: I think usually the AVX method for 128bit is just the same as the SSE one, as the Avx2 type literally inherits from SSE42

File: src/Maths/Silk.NET.Maths/Simd/Simd128.Divide.cs
Comment (HurricanKai) [inline src/Maths/Silk.NET.Maths/Simd/Simd128.Divide.cs:25]: Consider adding a TODO and raising an issue that non-floating point types need to be added here.

File: src/Maths/Silk.NET.Maths/Simd/Simd128.Multiply.cs
Comment (HurricanKai) [inline src/Maths/Silk.NET.Maths/Simd/Simd128.Multiply.cs:37]: Please add an SSE path (see https://stackoverflow.com/questions/8193601/sse-multiplication-16-x-uint8-t)

**Review by HurricanKai — CHANGES_REQUESTED** [review]
Reviewed SSE for now - I think the review comments apply to the other ones as well so not reviewing the rest

File: src/Maths/Silk.NET.Maths/Simd/Simd128.Abs.cs
Comment (WhiteBlackGoose) [reply src/Maths/Silk.NET.Maths/Simd/Simd128.Abs.cs:129]: But it doesn't exist 😢 

Hmm

Comment (HurricanKai) [reply src/Maths/Silk.NET.Maths/Simd/Simd128.Abs.cs:129]: I'm fine mering it if it doesn't compile 😛

File: src/Maths/Silk.NET.Maths/Simd/Simd128.Add.cs
Comment (WhiteBlackGoose) [reply src/Maths/Silk.NET.Maths/Simd/Simd128.Add.cs:49]: > This doesn't make sense - if AVX2 is supported, SSE2 is as well, so the lower branch is never hit - but I'm unsure whether we should prefer SSE or AVX here

Hmmmm. So I remove avx then?

Comment (HurricanKai) [reply src/Maths/Silk.NET.Maths/Simd/Simd128.Add.cs:49]: Yes

File: src/Maths/Silk.NET.Maths/Simd/Simd128.Multiply.cs
Comment (WhiteBlackGoose) [reply src/Maths/Silk.NET.Maths/Simd/Simd128.Multiply.cs:37]: May I leave a todo here too? To keep PRs not too heavy

Comment (HurricanKai) [reply src/Maths/Silk.NET.Maths/Simd/Simd128.Multiply.cs:37]: Sure

File: src/Maths/Silk.NET.Maths/Simd/Simd128.Add.cs
Comment (WhiteBlackGoose) [reply src/Maths/Silk.NET.Maths/Simd/Simd128.Add.cs:49]: I better remove sse

Comment (HurricanKai) [reply src/Maths/Silk.NET.Maths/Simd/Simd128.Add.cs:49]: II don't think that's better, SSE is available on NETCore 3.1, AVX isn't, I think

Comment (HurricanKai) [reply src/Maths/Silk.NET.Maths/Simd/Simd128.Add.cs:49]: Nevermind, but the check would still prevent some machines that can actually run this from running it, so SSE is preferable

File: src/Maths/Silk.NET.Maths/Simd/Simd128.Load.cs
Comment (Perksey) [inline src/Maths/Silk.NET.Maths/Simd/Simd128.Load.cs:311]: What if T is not 32-bit?

**Review by Perksey — COMMENTED** [review]
Not codeowner, so these are just general things I noticed in passing: kai can decide whether to care.

File: src/Maths/Silk.NET.Maths/Simd/Simd128.Divide.cs
Comment (Perksey) [reply src/Maths/Silk.NET.Maths/Simd/Simd128.Divide.cs:25]: Might be sensible to have a general SIMD tracking issue as there'll probably be lots of notes that we'll want to keep track of.

Comment (HurricanKai) [reply src/Maths/Silk.NET.Maths/Simd/Simd128.Divide.cs:25]: Seems fair to me as well

File: src/Maths/Silk.NET.Maths/Simd/Simd128.Load.cs
Comment (WhiteBlackGoose) [reply src/Maths/Silk.NET.Maths/Simd/Simd128.Load.cs:311]: Why does it matter?

Comment (Perksey) [reply src/Maths/Silk.NET.Maths/Simd/Simd128.Load.cs:311]: Because you'd get a segfault. This method is meant to load a pointer containing 4 Ts, not a pointer containing a 128-bit vector.

File: src/Maths/Silk.NET.Maths/Simd/Simd.GetWithElement.cs
Comment (WhiteBlackGoose) [reply src/Maths/Silk.NET.Maths/Simd/Simd.GetWithElement.cs:30]: This API existed before me, so I assumed there's no such 😅 
Will fix

Comment (WhiteBlackGoose) [reply src/Maths/Silk.NET.Maths/Simd/Simd.GetWithElement.cs:30]: Actually @HurricanKai was there a reason to have those methods, given that vector provides those methods?

Comment (HurricanKai) [reply src/Maths/Silk.NET.Maths/Simd/Simd.GetWithElement.cs:30]: Well, I don't remember why I did this, but I think something something JIT is confused.
I mean change it and check with disasmo, or try searching #lowlevel for this code

File: src/Maths/Silk.NET.Maths/Simd/Simd128.Add.cs
Comment (WhiteBlackGoose) [reply src/Maths/Silk.NET.Maths/Simd/Simd128.Add.cs:49]: Oh, no, I will remove avx for 128 bit

File: src/Maths/Silk.NET.Maths/Simd/Simd.GetWithElement.cs
Comment (Perksey) [reply src/Maths/Silk.NET.Maths/Simd/Simd.GetWithElement.cs:30]: These are aggressively inlined and do the same thing under-the-hood now, only it's validated and is `[Intrinsic]` (so the JIT may do special things if it feels like it!).

source: source.dot.net

Comment (HurricanKai) [reply src/Maths/Silk.NET.Maths/Simd/Simd.GetWithElement.cs:30]: I know them, but I remember something about it being wrong. As I said, try it.

File: src/Maths/Silk.NET.Maths/Simd/Simd128.Abs.cs
Comment (WhiteBlackGoose) [reply src/Maths/Silk.NET.Maths/Simd/Simd128.Abs.cs:121]: Uh. That would make your job of reviewer much more difficult

File: src/Maths/Silk.NET.Maths/Simd/Simd128.Load.cs
Comment (WhiteBlackGoose) [reply src/Maths/Silk.NET.Maths/Simd/Simd128.Load.cs]: Oh, right. It's more than that, it should convert `ptr` to `Vector128<T>*`, that's why

File: src/Maths/Silk.NET.Maths/Simd/Simd128.Abs.cs
Comment (WhiteBlackGoose) [reply src/Maths/Silk.NET.Maths/Simd/Simd128.Abs.cs:121]: I can probably add `And` as part of this PR too

File: src/Maths/Silk.NET.Maths/Simd/Simd.GetWithElement.cs
Comment (WhiteBlackGoose) [reply src/Maths/Silk.NET.Maths/Simd/Simd.GetWithElement.cs:30]: [Here](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABABgAJiBGAOgCUBXAOwwEt8YaBhCfAB1YAbGFADKIgG6swMXAG4AsACgylWoxbtOASRZRWTXNPnLV1eszYcaujPsPGaADQAcSRUtNUk5AGowwDGgAJgBWJAAeAwwAPnJxDFIMAAsAFQhSAH0AaWxWAAp/QJDwqJY4iQBKZXJa8gBeOIB1VhSAUWEOFnyJNHJSPtJKjy8fIqCoMMjouISktIzMgCEuABlCgImpstjyKpq6xr2aFvbOmG6B/uHlW6UAbQBZGBSIABNtfkF859ePr4A8nw2BBDDQAIIAc0hsFwRgkMF0ggMBkhlQAuqM/JsSpFUs1WskOjAuhgIviNsVJqV8XscVA+tFyAY3jAEH1UntsIIGDBKuQAO7JEQwcickDkZj4bBMbCQmBvZQAbwOtXGuPJcVhDEEGAadKpHjq5FgADMxeQYBR6iaYOaAKqGbCmzjg3ARdXUvExDkxfJm224HUYG5KY2O3DO11vN7+u2Wq4stn8m0Sbm8o11YgAdkDwY8AF8gA===) BCL's WithElement gives a better codegen than manually written. Also we probably should rely on BCL anyway?

Comment (WhiteBlackGoose) [reply src/Maths/Silk.NET.Maths/Simd/Simd.GetWithElement.cs:30]: [128 bit](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABABgAJiBGAOgCUBXAOwwEt8YaBhCfAB1YAbGFADKIgG6swMXAG4AsACgylWoxbtOASRZRWTXNPnLV1eszYcaujPsPGaADQAcSRUtNUk5AGowwDGgqACYXAB4DDAA+cnEMUgwACwAVCFIAfQBpbFYACn9A4LDIlliJAEplchryAF5YgHVWZIBRYQ4WPIk0clJe0gqPLx9CoKhQiKjY+MTU9IyAIS4AGQKA8cnSmPJK6tqG3Zpmto6YLv6+oeUbpQBtAFkYZIgAE21+QTynl/fPgHk+GwIIYaABBADmENguCMEhgukEBgMEIqAF0Rn4NsUIikmi0ku0YJ0MOE8esihMSnjdtioL0ouQDK8YAheildthBAwYBVyAB3JIiGDkDkgcjMfDYJjYCEwV7KADe+xqYxxZNiMIYggw9VplI8tXIsAAZqLyDAKHVjTAzQBVQzYE2cMG4cJqqm46Ls6J5U023DajDXJRGh24J0u16vP22i2XZmsvnWiRcnmG2rEADsAaDHgAvkA=) too, for 64-bit it's [about the same](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABABgAJiBGAOgCUBXAOwwEt8YaBhCfAB1YAbGFADKIgG6swMXAG4AsACgylWoxbtOASRZRWTXNPnLV1eszYcaujPsPGaADQAcSRUtNUk5AGowwDGhUAB4DDAA+cnEMUgwACwAVCFIAfQBpbFYACn9A4JQwliiJAEplckryAF4ogHVWBIBRYQ4WbIk0clIu0lKPLx88oKhQ8KiYuKSU1IAhLgAZXICRseLyMoqq2o2aBubWmHae7v7lc6UAbQBZGASIABNtfkFs2/unl4B5PjYIQxoAEEAObA2C4IwSGC6QQGAzA0oAXUGfhWBRCiXqjXiLRgbQwGIiy3yo0KmI2aKgXXC5AMDxgCC6iQ22EEDBgpXIAHd4iIYORmSByMx8NgmNhgTAHsoAN5bSrDdHk8EMQQYGoUkkeKrkWAAMwF5BgFGqupgBoAqoZsHrOIDcCFFaTCUyifqzbhVRgzkodVbcDa7Q8Htl3cbqUx6QhOaaJKz2dqqsQAOwer0eAC+QA===).

File: src/Maths/Silk.NET.Maths/Simd/Simd128.Load.cs
Comment (WhiteBlackGoose) [reply src/Maths/Silk.NET.Maths/Simd/Simd128.Load.cs]: @Perksey we're reading a 128-bit vector, so the bitness of T doesn't matter

File: src/Maths/Silk.NET.Maths/Simd/Simd.GetWithElement.cs
Comment (HurricanKai) [reply src/Maths/Silk.NET.Maths/Simd/Simd.GetWithElement.cs:30]: You need to look at the [loop version](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABABgAJiBGAOgCUBXAOwwEt8YaBhCfAB1YAbGFADKIgG6swMXAG4AsACgylWoxbtOASRZRWTXNPnLV1eszYcaujPsPGaADQAcSRUtNUk5AGowwDGgAJgBWJAAeAwwAPnJxDFIMAAsAFQhSAH0AaWxWAAp/QJDwqJY4iQBKZQBvZXIG8gAzaHJ86PJWcgBeclI5TvIIvwCgqDDI6JjuCEsB1gBqBcryCR7yAHVWFIBRYQ4WfIk0TpPSSo9GygB2VY8AX2UvHyKxibLY+JhElPSsgCEuAAZQqjEqTcqrapKOpKK4tKBtDpdXr9QbDV7gj7THhzTpLFZrXoSGhbXb7GCHVhnC71RrEW4SB5PJQAbQAst9khAACbafiCfKclK8/l8QQAeT4bAghhoAEEAOaK2C4IwSGC6QQGAyKyoAXWeI2K41KqTiZOSexgBwwEXNoJN73NqzBUBOyKYPJgCBOqVW2EEDBgKwA7skRDByP6QORmPhsExsIqYDzanSGpjTZEXaqGIIMOsNSbLo1YE1o+QYBReuXyABVQzYJqceW4CJZ50xP0xfJ1vMF2lwxqN3DN1s8nl9mAV6ser0+lbEwPB0sNBnkAcYB5AA)

It seems the check isn't elided in BCL

Comment (WhiteBlackGoose) [reply src/Maths/Silk.NET.Maths/Simd/Simd.GetWithElement.cs:30]: Indeed. But it looks more like a reason to PR to JIT than to replace with something which will be better for a short time. Have you filed an issue on that thing?

File: src/Maths/Silk.NET.Maths/Simd/Simd128.Abs.cs
Comment (WhiteBlackGoose) [reply src/Maths/Silk.NET.Maths/Simd/Simd128.Abs.cs:121]: Actually, as for consistency, wouldn't it make sense to leave it as it is? @HurricanKai . I can do `And` too, but that will require much more time as I will need to PR `Add` for Scalar before

File: src/Maths/Silk.NET.Maths/Simd/Simd.GetWithElement.cs
Comment (Perksey) [inline src/Maths/Silk.NET.Maths/Simd/Simd.GetWithElement.cs:25]: If we keep the manually written versions of these methods can we validate `index` please? This method is actually unsafe otherwise, rather than just (ab)using unsafe.

Comment (WhiteBlackGoose) [reply src/Maths/Silk.NET.Maths/Simd/Simd.GetWithElement.cs:25]: Then won't it destroy the purpose of the method? Kai's [comment](https://github.com/dotnet/Silk.NET/pull/665#discussion_r741953852). Actually, I personally would prefer to remove manually written methods, because it's not a good idea to have them in long-run perspective.

So can we decide one and forever here? @HurricanKai

File: src/Maths/Silk.NET.Maths/Simd/Simd128.Abs.cs
Comment (WhiteBlackGoose) [reply src/Maths/Silk.NET.Maths/Simd/Simd128.Abs.cs]: (`And` is added)

**Review by Perksey — CHANGES_REQUESTED** [review]
Will put my big X here until #668 is merged.

File: src/Maths/Silk.NET.Maths/Simd/Simd.GetWithElement.cs
Comment (HurricanKai) [reply src/Maths/Silk.NET.Maths/Simd/Simd.GetWithElement.cs]: I'm fine with removing them. We should for sure raise an issue on dotnet/runtime then, cause while not a huge deal and only used in fallback paths it'd be nice to have that anyways

File: src/Maths/Silk.NET.Maths/Simd/Simd128.Load.cs
Comment (HurricanKai) [reply src/Maths/Silk.NET.Maths/Simd/Simd128.Load.cs]: @Perksey this method doesn't read 4 Ts it reads Vector128<T>.Count Ts. All methods operate on that count

File: src/Maths/Silk.NET.Maths/Simd/Simd128.Constants.cs
Comment (HurricanKai) [inline src/Maths/Silk.NET.Maths/Simd/Simd128.Constants.cs:127]: nit: Scalar<T>.Zero  > default(T)

File: src/Maths/Silk.NET.Maths/Simd/Simd128.Load.cs
Comment (Perksey) [reply src/Maths/Silk.NET.Maths/Simd/Simd128.Load.cs]: @HurricanKai ok fair enough.

File: src/Maths/Silk.NET.Maths/Simd/Simd64.Constants.cs
Comment (HurricanKai) [inline src/Maths/Silk.NET.Maths/Simd/Simd64.Constants.cs:12]: Shouldn't there be an IsHardwareAccelerated somewhere in this file?

File: src/Maths/Silk.NET.Maths/Simd/Simd64.Not.cs
Comment (HurricanKai) [inline src/Maths/Silk.NET.Maths/Simd/Simd64.Not.cs:23]: Seems like we can make this whole method
```cs
if (Simd64.IsHardwareAccelerated) return Simd64.Xor(vector, Simd64<T>.AllBitsSet);
return Other(vec);
```

File: src/Maths/Silk.NET.Maths/Simd/Simd.GetWithElement.cs
Comment (Perksey) [reply src/Maths/Silk.NET.Maths/Simd/Simd.GetWithElement.cs]: I personally am not keen on not validating user input (segfaults are ugly), but can let this slide if Kai wants.

File: src/Maths/Silk.NET.Maths/Simd/Simd128.Not.cs
Comment (HurricanKai) [inline src/Maths/Silk.NET.Maths/Simd/Simd128.Not.cs:31]: Same as in Simd64.Not

File: src/Maths/Silk.NET.Maths/Simd/Simd256.Not.cs
Comment (HurricanKai) [inline src/Maths/Silk.NET.Maths/Simd/Simd256.Not.cs:31]: Same as in Simd64.Not

File: src/Maths/Silk.NET.Maths/Simd/Simd.GetWithElement.cs
Comment (HurricanKai) [inline src/Maths/Silk.NET.Maths/Simd/Simd.GetWithElement.cs:1]: Just putting this here so I don't forget whether we went through with this change

**Review by HurricanKai — CHANGES_REQUESTED** [review]

File: src/Maths/Silk.NET.Maths/Simd/Simd64.Constants.cs
Comment (WhiteBlackGoose) [reply src/Maths/Silk.NET.Maths/Simd/Simd64.Constants.cs:12]: @HurricanKai that's what I wanted to ask about. I didn't get what this method is supposed to do

File: src/Maths/Silk.NET.Maths/Simd/Simd64.Not.cs
Comment (WhiteBlackGoose) [reply src/Maths/Silk.NET.Maths/Simd/Simd64.Not.cs:23]: ._.

File: src/Maths/Silk.NET.Maths/Simd/Simd64.Constants.cs
Comment (HurricanKai) [reply src/Maths/Silk.NET.Maths/Simd/Simd64.Constants.cs:12]: Basically it's just a shorthand for "is one of the ISAs we use available" so usually `Sse(2).IsSupported || AdvSimd.IsSupported`

Comment (WhiteBlackGoose) [reply src/Maths/Silk.NET.Maths/Simd/Simd64.Constants.cs:12]: Got it

**Review by HurricanKai — APPROVED** [review]
I approve and thank you for your huge contribution!
@Perksey please remove your X

**Review by Perksey — APPROVED** [review]
eggs benedict

**Pull request merged**
